File: src/main/java/ru/arc/hooks/betterstructures/BSListener.java
package ru.arc.hooks.betterstructures;

import java.util.List;
import java.util.Set;
import java.util.concurrent.ConcurrentSkipListSet;

import com.magmaguy.betterstructures.api.BuildPlaceEvent;
import com.magmaguy.betterstructures.api.ChestFillEvent;
import com.magmaguy.betterstructures.config.generators.GeneratorConfigFields;
import org.bukkit.Chunk;
import org.bukkit.block.Block;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import ru.arc.bschests.PersonalLootManager;

import static com.magmaguy.betterstructures.config.generators.GeneratorConfigFields.StructureType.SURFACE;

public class BSListener implements Listener {

    Set<ChunkCoords> genCoords = new ConcurrentSkipListSet<>();
    int r = 7;

    @EventHandler
    public void onChestFill(ChestFillEvent event) {
        Block block = event.getContainer().getBlock();
        PersonalLootManager.processChestGen(block);
    }


    record ChunkCoords(int x, int z) implements Comparable<ChunkCoords> {
        @Override
        public int compareTo(ChunkCoords o) {
            return x == o.x ? Integer.compare(z, o.z) : Integer.compare(x, o.x);
        }
    }

    @EventHandler
    public void onStructureGen(BuildPlaceEvent event) {
        List<GeneratorConfigFields.StructureType> structureTypes = event.getFitAnything().getSchematicContainer().getGeneratorConfigFields().getStructureTypes();
        if (!structureTypes.contains(SURFACE)) {
            return;
        }
        if (event.getFitAnything().getLocation().y() < 64) {
            return;
        }

        Chunk chunk = event.getFitAnything().getLocation().clone().getChunk();
        ChunkCoords coords = new ChunkCoords(chunk.getX(), chunk.getZ());
        // check if chunk in radius r has already been generated
        for (int x = -r; x <= r; x++) {
            for (int z = -r; z <= r; z++) {
                if (genCoords.contains(new ChunkCoords(coords.x + x, coords.z + z))) {
                    event.setCancelled(true);
                }
            }
        }
        genCoords.add(coords);
    }

}

File: src/main/kotlin/ru/arc/hooks/betterstructures/BSListener.kt
package ru.arc.hooks.betterstructures

import com.magmaguy.betterstructures.api.BuildPlaceEvent
import com.magmaguy.betterstructures.api.ChestFillEvent
import com.magmaguy.betterstructures.config.generators.GeneratorConfigFields
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import ru.arc.bschests.PersonalLootModule
import java.util.concurrent.ConcurrentSkipListSet

class BSListener : Listener {

    private val genCoords = ConcurrentSkipListSet<ChunkCoords>()
    private val r = 7

    private data class ChunkCoords(val x: Int, val z: Int) : Comparable<ChunkCoords> {
        override fun compareTo(other: ChunkCoords): Int =
            if (x == other.x) z.compareTo(other.z) else x.compareTo(other.x)
    }

    @EventHandler
    fun onChestFill(event: ChestFillEvent) {
        PersonalLootModule.processChestGen(event.container.block)
    }

    @EventHandler
    fun onStructureGen(event: BuildPlaceEvent) {
        val structureTypes = event.fitAnything.schematicContainer.generatorConfigFields.structureTypes
        if (!structureTypes.contains(GeneratorConfigFields.StructureType.SURFACE)) return
        if (event.fitAnything.location.y() < 64) return

        val chunk = event.fitAnything.location.clone().chunk
        val coords = ChunkCoords(chunk.x, chunk.z)

        for (x in -r..r) {
            for (z in -r..r) {
                if (genCoords.contains(ChunkCoords(coords.x + x, coords.z + z))) {
                    event.isCancelled = true
                }
            }
        }
        genCoords.add(coords)
    }
}
